config/configHelpers.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+import merge from 'lodash.merge';
+import {
+  configurationType,
+  programmaticValue,
+  configurationLiteralValue,
+} from './types';
+
+/** helper function to get the value of an environment variable, or a defined fallback value */
+export const getEnvvarValue: (
+  name: string,
+  fallback?: configurationLiteralValue
+) => () => configurationLiteralValue = (
+  name,
+  fallback = `No value for ENVVAR ${name}`
+) => () => process.env[name] || fallback;
+
+/** helper used to process a `configurationDeclaration` or `featureFlagDeclaration` (D) to a `configurationType` with values being of type (T) for use across the codebase */
+export const processConfig: <D, T>(config: Array<D>) => configurationType<T> = (
+  config
+) => {
+  // merge all provided configs
+  const mergedConfig = config.reduce(
+    (acc, thisConfig) => merge(acc, thisConfig),
+    {}
+  );
+
+  // iterate through config
+  const processedConfiguration = Object.entries(mergedConfig).reduce(
+    (acc, [key, value]) => {
+      const valueType = typeof value;
+      const isLiteralValue =
+        valueType === 'string' ||
+        valueType === 'boolean' ||
+        valueType === 'number';
+      const isConfigurationValue =
+        !isLiteralValue &&
+        (value as programmaticValue<unknown, unknown>).configValue;
+
+      let publicValue, privateValue;
+
+      if (isLiteralValue) {
+        publicValue = value;
+        privateValue = value;
+      } else if (isConfigurationValue) {
+        const { configValue, publicConfigValue } = value as programmaticValue<
+          unknown,
+          unknown
+        >;
+        privateValue =
+          typeof configValue === 'function' ? configValue() : configValue;
+        publicValue = publicConfigValue;
+      } else {
+        const { values, publicValues } = processConfig([value]);
```

Comment:
Being a generic function - what happens if you supply a `D` that has children that don't match `programmaticValue`, `literalValue` or `D`? 

I'm (guessing) typescript will complain - but it means for e.g that if you passed in `{type: ["x"]}` it should either infinite loop or blow up.

Reply:
~It also seems to suggest that every returned config value is of type `T` (so I'm guessing you then just have to pass in `T` as `unknown` or a subset.~

Saw later down how you can pass it through. I think the generic does work, but maybe the comment needs a bit more clarity to suggest that every value in `D` needs to be of type `T`

Reply:
It almost suggests that `D` needs `T` as a parameter somehow too (like `<D, T>(config: Array<D<T>>) => configurationType<T>`)

Reply:
> Being a generic function - what happens if you supply a `D` that has children that don't match `programmaticValue`, `literalValue` or `D`?
> 
> I'm (guessing) typescript will complain - but it means for e.g that if you passed in `{type: ["x"]}` it should either infinite loop or blow up.

It will - and with the config we have, the code wont compile without it being the right type (without config changes).

Will look to update typing now - when used I do specify the exact D and T https://github.com/matthew-chirgwin/strimzi-ui/blob/f3feb7ea63fabb60d9ee7e47d11de13e1b25f5f4/config/index.ts#L18 - but agree reading the function in isolation will make this hard to follow